packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,310 @@ interface Window {
+
+(function() {
+
+/**
+ * Describes the state machine for loading a single module.
+ *
+ * At runtime, these are integers that are inlined in their use sites.
+ */
+const enum StateEnum {
+  /**
+   * The initial state.
+   */
+  Initialized,
+
+  /**
+   * Comes after Initialized. We have begun loading the module over the network.
+   * Toplevel scripts skip this state entirely.
+   */
+  Loading,
+
+  /**
+   * Comes after Loading. The module's <script> tag has loaded. If there was a
+   * define() call for this module, it has run by now. We have started loading
+   * the module's dependencies, but we can't execute them until some other,
+   * earlier scripts have executed.
+   */
+  WaitingOnEarlierScripts,
+
+  /**
+   * Comes after WaitingOnEarlierScripts. All earlier scripts are now executed,
+   * and we can now execute our dependencies in order. Once that's done, we can
+   * execute this module too.
+   */
+  WaitingOnDeps,
+
+  /**
+   * The successful terminal state. Comes after WaitingOnDeps. All of the
+   * module's dependencies have loaded and executed, and the module's body, if
+   * any, has executed, and there were no errors.
+   */
+  Executed,
+
+  /**
+   * The unsuccessful terminal state. Something went wrong, either leading up
+   * to, or during execution of the module.
+   */
+  Failed,
+}
 
-interface Module {
+interface Initialized {
+  state: StateEnum.Initialized;
+}
+interface Loading {
+  state: StateEnum.Loading;
+}
+interface BaseWaiting {
+  /**
+   * The dependencies of this module, in order.
+   */
+  deps: Module[];
+  /**
+   * Args that we will pass into to moduleBody.
+   */
+  args: Array<{}>;
+  /**
+   * The body of the module, which is executed after its dependencies are
+   * loaded.
+   *
+   * In AMD/Commonjs terminology, this is the factory function.
+   */
+  moduleBody: undefined|Function;
+}
+interface WaitingOnEarlierScripts extends BaseWaiting {
+  state: StateEnum.WaitingOnEarlierScripts;
+}
+interface WaitingOnDeps extends BaseWaiting {
+  state: StateEnum.WaitingOnDeps;
+}
+interface Executed {
+  state: StateEnum.Executed;
+}
+interface Failed {
+  state: StateEnum.Failed;
+  readonly error: Error;
+}
+type ModuleStateData =
+    Initialized|Loading|WaitingOnEarlierScripts|WaitingOnDeps|Executed|Failed;
+
+interface Module<SD extends ModuleStateData = ModuleStateData> {
   url: NormalizedUrl;
   urlBase: NormalizedUrl;
   exports: {[id: string]: {}};
-  /** This is a top-level module. */
+  stateData: SD;
+  /** True if this is a top-level module. */
   isTopLevel: boolean;
-  /** This module hasn't started loading its script yet. */
-  needsLoad: boolean;
-  /** All dependencies are resolved and the factory has run. */
-  resolved: boolean;
-  error: Error|undefined;
-  /** Callbacks from dependents waiting for this module to resolve. */
-  onResolve: Array<() => void>;
-  onError: Array<ErrorCallback>;
+  /**
+   * Callbacks that are called exactly once, for the next time the module
```

Comment:
What kind of callbacks? Is there a named type that can describe this here and be searchable in the rest of the module?

Reply:
We only use this in two places, and since this is a zero argument function that returns nothing I'd say just look for `onNextStateChange`.